### ISSUE TITLE
EIP-1559 Transactions

### ISSUE THREAD

**Issue by unparalleled-js** [mikeshultz/ledger-eth-lib#9]
Hello!

This is a feature request for EIP-1559 support.

Thank you for this repo, it has been a nice resource to look at.

I have been trying implement this myself using the `eth-account` library in another project but am running into issues

**Comment by mikeshultz** [comment]
Looks like it should be possible as [ledger app-ethereum appears to added support some months ago](https://github.com/LedgerHQ/app-ethereum/pull/157).  I'll poke around, maybe this weekend.

**Comment by unparalleled-js** [comment]
Yes I can get it to work when I already have the encoded transaction.
It is just the encoding of the transaction I am struggling with

**Comment by mikeshultz** [comment]
Just some notes as I'm going through this.

Encoding seems to be a bit different.  [EIP-2718](https://eips.exposed/eip/2718) defines typed transactions.  There's a handful of different unsigned encoded transaction types:

- Original/Legacy - `rlp([nonce, gasprice, startgas, to, value, data, v, r, s])`
- EIP-155 (Same as legacy with chain ID) - `rlp([nonce, gasprice, startgas, to, value, data, chainid, 0, 0])`
- Type 1 transactions ([EIP-2930](https://eips.exposed/eip/2930) - optional access lists) - `1 || rlp([chainId, nonce, gasprice, startgas, to, value, data, access_list])`
- Type 2 transactions ([EIP-1559](https://eips.exposed/eip/1559)) - TBD

The encoding of type 2 does not seem to appear in EIP-1559, so I'm still digging that up.  I'll probably want to add Type 1 transactions to this as well, presuming they work with Ledger's `app-ethereum`.

**Comment by mikeshultz** [comment]
The linked page for EIP-1559 is kind of out of date(oops).  [Actual encoding should be like this](https://eips.ethereum.org/EIPS/eip-1559) for type 2 transactions:
 
    0x02 || rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, destination, amount, data, access_list])

Testing this encoding out, but I'm having some issues getting this to work.  Ledger seems to think the data is out of order.  It does look like it expects `TransactionType` to be packed with typed transactions which I'm doing:

https://github.com/LedgerHQ/app-ethereum/blob/12ebced7bf65f0aa808f3afec73aa40ad80aeff9/src_features/signTx/cmd_signTx.c#L47

I'm not super great with C though and have no idea what this is:

https://github.com/LedgerHQ/app-ethereum/blob/12ebced7bf65f0aa808f3afec73aa40ad80aeff9/src_features/signTx/cmd_signTx.c#L51

Looking through how it processes fields, they seem to be in the order of this enum (since it increments a field index as it processes the bytes:

https://github.com/LedgerHQ/app-ethereum/blob/f0f47e4250430edce99967742d9a16c9ee8bd29c/src_common/ethUstream.h#L81-L95

Which looks like it matches the definition in the EIP.  So, I have something I think should work, but it doesn't.  Need to verify my code a bit and will try some things out.  Was hoping this would be a quick update but guess not.  Will update if I learn more.

**Comment by mikeshultz** [comment]
Success!  

I was able to decode the tx they're using in their test case (which I think is actually invalid, anyway because the bribe and max fee were swapped) and use that to figure out what was wrong with my implementation.  Looks like I wasn't RLP encoding `access_list` properly.  I now have a successful implementation and I've been able to sign a valid transaction.  I'm kind of crashing however, and probably won't be wrapping it up tonight.  I still want to do a bit more testing to be sure this works and then clean up my quick & dirty implementation.  Should probably be able to take care of that and probably release tomorrow night.
